x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java

```diff
@@ -518,8 +529,14 @@ private void putComposableTemplate(
                 new ActionListener<AcknowledgedResponse>() {
                     @Override
                     public void onResponse(AcknowledgedResponse response) {
-                        creationCheck.set(false);
-                        if (response.isAcknowledged() == false) {
+                        if (response.isAcknowledged()) {
+                            if (isUpgrade && applyRolloverAfterTemplateV2Upgrade()) {
+                                invokeRollover(state, templateName, indexTemplate, creationCheck);
```

Comment:
Not sure if I was very clear when we chatted yesterday, so just reiterating in case: we need to consider that a failure may occur between the template update and rollover. If we don't cater for that, we'll end up with data streams in a half-updated state, and would require manual rollovers to resolve.

Could we make `invokeRollover` idempotent by comparing the active data stream backing index's applied template version to the current template version? Then we should be able to check if a rollover is needed on cluster state change, and be more resilient to failures.

If I'm reading the code properly, that's the approach taken in the profiling plugin: https://github.com/elastic/elasticsearch/blob/d1a2848bb3f69bb9990aa38fee99a544730efe75/x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/AbstractProfilingPersistenceManager.java#L75-L117

(Specifically see the `getIndexState` code.)

CC @danielmitterdorfer. We should probably align approaches so the profiling plugin could use this general logic in the future. WDYT?

Reply:
> we need to consider that a failure may occur between the template update and rollover. If we don't cater for that, we'll end up with data streams in a half-updated state, and would require manual rollovers to resolve.

Note that we invoke a rollover only if the update is [acknowledged](https://github.com/elastic/elasticsearch/pull/97546/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R532) by the cluster, which should guard from that IIANM.

> Could we make `invokeRollover` idempotent by comparing the active data stream backing index's applied template version to the current template version?

If I do that, I lose the existing synchronization mechanism that ensures that multiple updates of a template cannot happen concurrently. Follow the [`creationCheck`](https://github.com/elastic/elasticsearch/pull/97546/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R516) and see how I utilized this mechanism so that index template updates and the corresponding rollovers happen atomically (in the context of the registry, not within the cluster).

Reply:
> CC @danielmitterdorfer. We should probably align approaches so the profiling plugin could use this general logic in the future. WDYT?

Thanks for the ping. In the profiling plugin we have chosen a different approach that decouples versioning of index templates from matching data streams (and indices) and their versioning. We do this because: 

* Not all changes to the index template require a rollover (e.g. removing a property in the template does not require a rollover).
* In some cases (e.g. for key-value lookup indices) rollovers are disruptive to cluster performance and disk usage.

Therefore we use `IndexTemplateRegistry` only to manage index templates (and ILM policies) and have introduced separate management classes (`ProfilingDataStreamManager` and `ProfilingIndexManager`) to manage the lifecycle of data streams / indices. These classes are inspired by `SystemIndexMappingUpdateService` that manages mapping updates for system indices. I wonder whether separating these two concerns (with some common infrastructure in the core plugin) would also make sense here?

Reply:
What we should really avoid is that every team solves the same problems in slightly different ways. As much as possible, we should try to contribute the enhancements from the APM and profiling plugins to the Elasticsearch core plugin.

Reply:
I think it's reasonable to defer an aligned solution in favor of iterating on different approaches up to a certain point. However, as these different solutions become more sophisticated, it will become increasingly complex to come up with a solution that satisfies all needs.

In this case, the profiling solution is already quite sophisticated (for example with migrations) and the apm solution doesn't seem to be adding lots of complexity. Also, none of the different solutions really expose something to users or have backwards compatibility guarantees, other than setting requirements on minimizing the number of rollovers. So I think I'm comfortable with moving along with something APM specific and deferring consolidation for a bit longer under the condition that the ES, APM, and the profiling teams commit to working on a common solution.

The first step of that would be to list the high-level requirements each team has to see if they're compatible and to evaluate the feasibility.

Reply:
Just clarifying - what I added is indeed quite specific for the current basic APM requirements, but it is added as a general capability for any registry to opt-in to incur rollovers automatically upon index template upgrade. Probably not a functionality we will want to maintain once we have https://github.com/elastic/elasticsearch/issues/96521 or the like.

It is as simple as it gets, both on the positive and the negative side. The negative is of course that any registry update will trigger rollovers all over. Besides the obvious overhead related to that, it will probably also cause lots of excessive cluster state change events to be evaluated. Not sure this is an issue in itself, but worth being aware.